### DISCUSSION TITLE
MATCH_RECOGNIZE keyword for pattern matching

### DISCUSSION THREAD

**Discussion by david542542** [duckdb/duckdb#3994] [Ideas]
The `MATCH_RECOGNIZE` clause is an incredibly powerful way to do pattern matching in SQL. For example, here is a description of its implementation in [Snowflake](https://docs.snowflake.com/en/sql-reference/constructs/match_recognize.html) or [Azure Stream Analytics](https://docs.microsoft.com/en-us/stream-analytics-query/match-recognize-stream-analytics). 

The only open source implementation I've come across is [Apache Flink](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/sql/queries/match_recognize/) (Github [here](https://github.com/apache/flink/search?l=Java&q=match_recognize)). It'd be quite a challenging undertaking but an awesome feature to have in DuckDB!

**Comment by bo5o** [comment]
+1 for MATCH_RECOGNIZE (aka Row Pattern Matching)

There is another open-source implementation in Trino (formerly known as PrestoSQL). It has very good [documentation](https://trino.io/docs/current/sql/match-recognize.html) and you can find the code on [github](https://github.com/trinodb/trino/commit/e41d9af7a8925336a2c7b1ae227059395cffdd49).

**Reply by jimdickson007** [reply]
I believe neither Trino nor Flink implement backtracking in MATCH_RECOGNIZE.
Oracle and Snowflake both do - and it makes it much more powerful.

**Comment by voberoi** [comment]
+1 for MATCH_RECOGNIZE

**Comment by zhaoyongjie** [comment]
+1

**Comment by soerenwolfers** [comment]
Actual "regular" expressions (in the sense of detecting a regular language, via finite state machines) can be detected using the `reduce` window function proposed at  https://github.com/duckdb/duckdb/discussions/10703 ; everything else isn't guaranteed to finish in linear time and is therefore less attractive (or at least should arguably be discussed separately).

For example, to detect the pattern "price below 1 on day X, price above 1 between X and Y, price below 1 on day Y" you could use a three-state state machine as such

```sql
SELECT
  reduce(1, state -> state_machine(state, 
    [
      [price >= 1, price < 1, false, false], # Move from state 1 to 2 if a <1 price is detected 
      [false, price < 1, price >= 1, false], # Stay if <1; move on if >=1
      [false, false, price >= 1, price < 1], # Stay if >= 1; conclude if < 1
      [false, price < 1, price >= 1, false]
    ], # Final state has same transitions as state 2 because final condition is simultaneously start of new match
  ))
  OVER (PARTITION BY product ORDER BY date)
FROM df
```

where 

```sql
CREATE MACRO state_machine(state, transitions) AS 
CASE
  WHEN transitions[state][1] THEN 1
  WHEN transitions[state][2] THEN 2
  WHEN transitions[state][3] THEN 3
  WHEN transitions[state][4] THEN 4
  ... --- can't think of a way right now to implement this non-templated, but that would be easy to offer in C++
END
```

That being said: I completely acknowledge that a regular expression would look better than this (even if it too were limited to FSM), but maybe the `reduce` functionality could still be a sensible stepping stone towards the full `MATCH_RECOGNIZE`.

**Comment by Maxxen** [comment]
Just CC:ing @taniabogatsch as she actually wrote her bachelor thesis on how to efficiently implement match-recognize on modern hardware. Not sure we're planning to add this is anytime soon though.

**Reply by jimdickson007** [reply]
A subjective observation -
A lot of the problems I attack lie in intersection of graph-style traversal and relational data structures.
Think Travelling Salesman et al
So I am always on the lookout for a database that can do both.
Duck, Umbra/CedarDB, Kuzu seem to be primed to be in this space.
Alternatively, Oracle and maybe Snowflake.

Addition of MATCH_RECOGNIZE / GQL could make Duck the graph db killer.

reference : https://github.com/duckdb/duckdb/discussions/5933

**Reply by Maxxen** [reply]
@jimdickson007 you might be interested in https://github.com/cwida/duckdb-pgq and https://github.com/cwida/duckpgq-extension

**Reply by jimdickson007** [reply]
Also noticed it is on the roadmap - not sure if that is news or not
https://duckdb.org/roadmap

**Comment by dridk** [comment]
Hi, 

Clickhouse has a nice grammar : https://clickhouse.com/docs/sql-reference/aggregate-functions/parametric-functions#sequencematch 

```
SELECT sequenceCount('(?1)(?t>1800)(?2)')(time, condition1, condition2) FROM events

```

**Reply by vkingnew** [reply]
this feature in which vesion will be released general available ？

**Comment by jimdickson007** [comment]
Have not tried them myself yet but these are interesting

https://github.com/Query-farm/vgi-matchrecognize/blob/main/README.md

https://altertable.ai/blog/2026-08-07-match-recognize-that-scales

**Comment by redox** [comment]
We've indeed built one at https://altertable.ai/ and currently testing it with beta-testers, shoot us an email if you want to try it. And yes thanks for sharing @jimdickson007, we indeed blogged about it here: https://altertable.ai/blog/2026-08-07-match-recognize-that-scales

**Comment by hannes** [comment]
PR is here https://github.com/duckdb/duckdb/pull/25255

**Comment by hannes** [comment]
This is now in `main`. We plan to release `MATCH_RECOGNIZE` support with DuckDB 2.1 (not the upcoming 2.0). Feel free to try it out on a nightly build. 

Here is a fun example:
```SQL
-- old and busted
SELECT '..↑↑↓↓←→←→BA..' ~ '↑{2}↓{2}(←→){2}BA' konami_found;

-- new hotness
FROM (
PIVOT (
	FROM UNNEST('..↑↑↓↓←→←→BA..'.split('')) 
		WITH ORDINALITY t(c,i)) 
ON c GROUP BY i) 
MATCH_RECOGNIZE (
	DEFINES AUTO
	ORDER BY i 
	PATTERN (↑{2} ↓{2} (← →){2} B A) 
	MEASURES true AS konami_found);
```
